public/typescript/authentication.ts

```diff
@@ -31,8 +31,9 @@ declare const publicKeyOptions: SerializedPublicKeyCredentialRequestOptions;
     updateState(S.REQUEST_USER_FOR_ASSERTION),
     requestUserAssertion,
     updateState(S.PUBLIC_KEY_CREDENTIALS),
-    concatMap((credentials) => isPublicKeyCredentialType(credentials) ?
-      of(credentials).pipe(
+    concatIfElse(
```

Comment:
**A general authentication / registration feedback**

Let me start with stating that I'm by no means an expert in the realm of functional programming. Having said that. The code provided in the registration and authentication scripts are rather complex. 

And for a casual reader, are not very expressive. For one, use of domain specific functions like `updateState`, `whenResponseIsOk`, `requestUserAssertion` are used intertwined with RXJS functions like `shareReplay`, `retryWhen` and others. This forces the reader to have knowledge of these infrastructure specific tools, in order to understand what is happening on the 'application' level. Maybe it's possible to wrap those, enabling us to use more meaningful naming. As an alternative, inline documentation could be added, explaining why certain less obvious actions are taken.

I'd love to see the complexity brought down (if possible), hopefully making the logic more easy to understand and, more importantly, to maintain by developers less experienced in the functional programming realm.